File: Ch03/03_03/stats.py
class StatsError(Exception):
    pass


def mean(values):
    if not values:
        raise StatsError('"mean" with no values')
    return sum(values) / len(values)


if __name__ == '__main__':
    values = [2, 4, 8]
    try:
        print(mean(values))
    except StatsError as err:
        print(f'error: {err}')

File: Ch03/03_03/stats.go
package main

import "fmt"

func Sum(values []float64) float64 {
	total := 0.0
	for _, v := range values {
		total += v
	}
	return total
}

func Mean(values []float64) (float64, error) {
	if len(values) == 0 {
		return 0, fmt.Errorf("Mean of empty slice")
	}

	return Sum(values) / float64(len(values)), nil
}

func main() {
	values := []float64{2, 4, 8}
	m, err := Mean(values)
	if err != nil {
		fmt.Printf("error: %s\n", err)
		return
	}
	fmt.Println(m)
}
